src/main/cli/cli.c

```diff
@@ -3745,10 +3756,54 @@ static void cliDumpRateProfile(uint8_t rateProfileIndex, uint8_t dumpMask)
     rateProfileIndexToUse = CURRENT_PROFILE_INDEX;
 }
 
+#ifdef USE_CLI_BATCH
+static void cliPrintCommandBatchWarning(const char *warning)
+{
+    cliPrintErrorLinef("Errors were detected - please review before continuing");
+    if (warning) {
+        cliPrintErrorLinef(warning);
+    }
+}
+
+static void resetCommandBatch(void)
+{
+    commandBatchActive = false;
+    commandBatchError = false;
+}
+
+static void cliBatch(char *cmdline)
+{
+    if (strncasecmp(cmdline, "start", 5) == 0) {
+        if (!commandBatchActive) {
+            commandBatchActive = true;
+            commandBatchError = false;
+        }
+        cliPrintLine("Command batch started");
+    } else if (strncasecmp(cmdline, "end", 3) == 0) {
+        if (commandBatchActive && commandBatchError) {
+            cliPrintCommandBatchWarning(NULL);
+        } else {
+            cliPrintLine("Command batch ended");
+        }
+        resetCommandBatch();
+    } else {
+        cliPrintErrorLinef("Invalid option");
+    }
+}
+#endif
+
 static void cliSave(char *cmdline)
 {
     UNUSED(cmdline);
 
+#ifdef USE_CLI_BATCH
+    if (commandBatchActive && commandBatchError) {
```

Comment:
I still don't like that this makes the `save` command context sensitive - it would be much cleaner if a `save force` or similar was required to persist if there were errors, not just a repetition of the same command until the board resets.

Reply:
I still think this is very intuitive. Remember that users generally won't be using the `batch` command or even really notice it. So it's not like they will cause this workflow with normal CLI activity. The only time this change in `save` behavior will occur is if the user pastes in a `diff` (or `dump`) and then we're giving a courtesy "heads-up" that something was wrong and they should probably look through the output because it would have scrolled off or been cleared by the save. When the user is entering commands on their own the error should be visible and they can fix as appropriate.

The other argument is what about if the user pastes in a block of commands that didn't come from a `diff` and thus didn't get the `batch` commands added? Well nothing will change in that case. The original solution addressed that by having the Configurator detect the "pasting" event and auto-wrapping the block but we decided not to go that direction.

Reply:
I think that the approach to have modifiers on commands to have them behave differently is well understood by users, just like `diff` vs. `diff all`.

One more reason for not wanting to use identical commands, even more so in the case of copy / pasting backups is that, depending on whether the user pastes the backup with or without the `save` at the end, manually entering save will either save, or force save, if identical commands are used for both of these.